holmes/plugins/toolsets/kubernetes_logs.py

```diff
@@ -0,0 +1,284 @@
+                    params=params, previous=True
+                ) + self._fetch_pod_logs(params=params, previous=False)
+            else:
+                all_logs = self._fetch_pod_logs(
+                    params=params, containers=pod.containers, previous=True
+                )
+                all_logs = all_logs + self._fetch_pod_logs(
+                    params=params, containers=pod.containers, previous=False
+                )
+
+            if params.filter_pattern:
+                all_logs = self._filter_logs(all_logs, params.filter_pattern)
+
+            if params.limit and params.limit < len(all_logs):
+                all_logs = all_logs[-params.limit :]
+
+            # Format output
+            formatted_logs = self._format_logs(all_logs)
+
+            return StructuredToolResult(
+                status=ToolResultStatus.SUCCESS,
+                data=formatted_logs,
+                params=params.model_dump(),
+            )
+        except ApiException as e:
+            return StructuredToolResult(
+                status=ToolResultStatus.ERROR,
+                error=f"Kubernetes API error: {str(e)}",
+            )
+        except Exception as e:
+            logging.exception(f"Error fetching logs for pod {params.pod_name}")
+            return StructuredToolResult(
+                status=ToolResultStatus.ERROR,
+                error=f"Error fetching logs: {str(e)}",
+            )
+
+    def _find_pod(self, namespace: str, pod_name: str) -> Optional[Pod]:
+        if not self._core_v1_api:
+            return None
+        try:
+            pod: Any = self._core_v1_api.read_namespaced_pod(
+                name=pod_name, namespace=namespace
+            )
+            return Pod(containers=[container.name for container in pod.spec.containers])
+        except ApiException as e:
+            if e.status == 404:
+                return None
+            else:
+                raise
+        except Exception:
+            logging.exception(f"Error getting pod {pod_name}", exc_info=True)
+            raise
+
+    def _fetch_pod_logs(
+        self,
+        params: FetchLogsParams,
+        containers: Optional[List[str]] = None,
+        previous: bool = False,
+    ) -> List[Any]:
+        pod_logs = []
+        if containers and len(containers) > 1:
+            # Fetch logs for each container, try current logs first then previous if none found
+            for container_name in containers:
+                container_logs = self._fetch_logs(
+                    params=params,
+                    container_name=container_name,
+                    previous=previous,
+                )
+
+                # Add container name prefix only for multi-container pods
+                # This matches kubectl behavior
+                container_logs = [f"{container_name}: {log}" for log in container_logs]
+                pod_logs.extend(container_logs)
+
+        if not pod_logs:
```

Comment:
Just to make sure that it is be design, if we try to call fetch_logs for each container and none of them return logs because no logs exists we still call this? Right?
If this is indeed the case can you elaborate the use case for it?

Reply:
Good question, I added a comment to explain it.
It's a way to be resilient, in particular for previous logs. The container list is a 'current' container list (as far as I know I can't get a 'previous container' list). This means that fetching previous logs with specific containers will work in most cases but may fail to fetch logs if the container names have changed